capsules/src/mac.rs

```diff
@@ -0,0 +1,365 @@
+//! Implements 802.15.4 MAC device functionality as an abstraction layer over
+//! the raw radio transceiver hardware. The abstraction difference between a MAC
+//! device and a raw radio transceiver is that the MAC devices exposes a
+//! frame-oriented interface to its users, whereas the radio transceiver
+//! transmits raw byte sequences. There is some abstraction breaking here,
+//! though because the following are still implemented at the hardware level:
+//! - CSMA-CA backoff
+//! - FCS generation and verification
+//!
+//! TODO: Encryption/decryption
+//! TODO: Sending beacon frames
+//! TODO: Channel scanning
+
+use core::cell::Cell;
+use kernel::ReturnCode;
+use kernel::common::take_cell::TakeCell;
+use kernel::hil::radio;
+use net::ieee802154::*;
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub struct FrameInfo {
+    // These offsets are relative to buf[radio::PSDU_OFFSET..] so that
+    // the mac frame length is data_offset + data_len
+    mac_payload_offset: usize,
+    data_offset: usize,
+    data_len: usize,
+    security_params: Option<(Security, [u8; 16])>,
+}
+
+impl FrameInfo {
+    pub fn frame_length(&self) -> usize {
+        self.data_offset + self.data_len
+    }
+
+    pub fn remaining_data_capacity(&self, buf: &[u8]) -> usize {
+        buf.len() - radio::PSDU_OFFSET - radio::MFR_SIZE - self.data_offset - self.data_len
+    }
+
+    pub fn append_payload(&mut self, buf: &mut [u8], payload: &[u8]) -> ReturnCode {
```

Comment:
OK, this the central method of `FrameInfo`, so I'll ask here.


Why not internalize the `buf` into `FrameInfo`, i.e. when a `FrameInfo` is created, it stores the buffer as a field. Then, `append_payload` would just look like this:

```rust
pub fn append_payload(&mut self, payload: &[u8]);
```

I think this would solve a couple issues with the current design:

  1. Currently, the user should make sure to call `prepare_data_frame` and `transmit` with the same `buf`, but nothing is enforcing that.

  2. Currently, I believe, the user could modify the header that `prepare_header` serializes into the buffer before calling `transmit`. That's _maybe_ OK from a safety perspective (I haven't thought about whether anything bad can happen), but it seems simpler if we can just assume the user hasn't touched the header part of `buf` in between calls.

If you do this, it probably makes sense to rename this just to `Frame` or something.

It's also pretty possible I've got this wrong. _But_ if you keep the interface as it is, some documentation somewhere should probably help someone (like me!) understand why.